services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/IdentityTemplate.java

```diff
@@ -0,0 +1,152 @@
+         */
+        String getPlaceHolder() {
+            return placeHolder;
+        }
+
+        /**
+         * Extracts the attribute value from the given list of RDNs (Representation of Distinguished Names).
+         * <p>
+         * If the RDNs list contains multiple occurrences of the same attribute, the value
+         * of the first occurrence is returned.
+         *
+         * @param rdns The list of RDNs.
+         * @return The attribute value or {@code null} if the attribute is not available in the RDNs list.
+         * @throws NullPointerException if rdns is {@code null}.
+         */
+        String getValue(final List<Rdn> rdns) {
+            Objects.requireNonNull(rdns, "rdns list must not be null");
+
+            return rdns
+                    .stream()
+                    .filter(rdn -> this.toString().equalsIgnoreCase(rdn.getType()))
+                    .findFirst()
+                    .map(Rdn::getValue)
+                    .filter(String.class::isInstance)
+                    .map(String.class::cast)
+                    .orElse(null);
+        }
+    }
+
+    /**
+     * Applies attribute values from the given subject DN to the template.
+     * <p>
+     * The following placeholders are supported.
+     * <ul>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_DN} for <em>Subject Distinguished Name (DN)</em></li>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_CN} for <em>Common Name (CN)</em></li>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_OU} for <em>Organizational Unit Name (OU)</em></li>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_O} for <em>Organization Name (O)</em></li>
+     * </ul>
+     *
+     * @param subjectDN The subject DN.
+     * @return The filled template.
+     * @throws IllegalArgumentException if the subject DN is not valid or any of the attributes
+     *                                  configured in the template are not present in the subject DN.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public String apply(final String subjectDN) {
+        Objects.requireNonNull(subjectDN, "subjectDN must not be null");
+
+        try {
+            final List<Rdn> rdns = new LdapName(subjectDN).getRdns();
+            String result = template.replaceAll(QUOTED_PLACEHOLDER_SUBJECT_DN, subjectDN);
+            for (final Attribute attribute : Attribute.values()) {
+                result = applyAttribute(attribute, result, rdns);
+            }
+            return result;
+        } catch (final InvalidNameException e) {
+            throw new IllegalArgumentException(String.format("subject DN [%s] is not valid", subjectDN));
+        }
+    }
+
+    private static String applyAttribute(final Attribute attribute, final String template,
+            final List<Rdn> rdns) {
+        if (template.contains(attribute.getPlaceHolder())) {
```

Comment:
the content of this conditional block is fine. My point is that we can simply run it unconditionally because the `template.replaceAll()` method will perform the same check, i.e. it will only replace occurrences of the attribute's corresponding placeholder if the placeholder is actually contained in the template, won't it?

Reply:
The `IllegalArgumentException` is to be thrown only when a placeholder for an attribute is configured in the template and also that attribute is not found in the subejct DN. If we remove that conditional statement, then an `IllegalArgumentException` will be thrown whenever an attribute is not found in _SubjectDN_ irrespective of  the corresponding placeholder is configured in the template or not. Hence we need that check. Sorry, I didn't get it how it can be achieved only with `template.replaceAll()`.

Reply:
right, my bad